VnodeCacheDesign.md

```diff
@@ -0,0 +1,55 @@
+- **TBD:** Eliminate the use of file flags to determine if a file is inside of a root
+
+## Design 
+
+### Cached Data
+
+PrjFSKext will cache the following information for each vnode:
+
+- The `vnode_t` itself
+- The vnode's `uint32_t` vid (generation number)
+- `VirtualizationRootHandle` for the vnode (or `RootHandle_None` if the vnode is not in a root).  **NOTE** This assumes that `VirtualizationRootHandle`  are not recycled as VFS4G instances are unmounted\remounted.
+
+The data for each vnode will be stored in a hash table that assumes there will be `desiredvnodes` total vnodes.
+
+### Using the Cache
+
+When PrjFSKext needs to determine the virtualization root of a file (or if a file is inside of a virtualization root at all), it will look up the vnode in its cache.
+
+- If the vnode exists in the cache and its `vid` has not changed, PrjFSKext will use the `VirtualizationRootHandle` in the cache.
+
+- If the vnode exists in the cache and its `vid` has changed, PrjFSKext will lookup its `VirtualizationRootHandle` (using the existing code in PrjFSKext) and update the cache entry.
+
+- If the vnode does not exist in the cache,  PrjFSKext will lookup its `VirtualizationRootHandle` (using the existing code in PrjFSKext) and insert the vnode into the cache.
+
+### Special Case: Renames
+
+*Directory renames* 
+
+Directory renames are a special case because all of the directory's children may need to have their `VirtualizationRootHandle` updated. Rather than recording\tracking all of the parent\child relationships among the vnodes PrjFSKext will simply invalidate the entire cache when it detects that a directory has been renamed in\out of a virtualization root.  If PrjFSKext is unable to determine if the rename was within the same root (e.g. the directory's vnode was recycled) then PrjFSKext will invalidate the cache.
+
+*File renames*
+
+When a file is renamed PrjFSKext will update its `VirtualizationRootHandle` and\or `vid` as needed.
+
+## Future Enhancements
+
+- The vnode cache could track which vnodes have received the `KAUTH_FILEOP_WILL_RENAME` file operation (on Mojave) so that when handling `KAUTH_VNODE_DELETE` the kext can know if the delete vnode operation is for a rename.
+
+## Issues not addressed by the cache
+
+- Improving the speed and reliability of finding a vnode's path
+- Addressing failures to get a vnode's parent (when the vnode is not in the cache)
+- Finding the roots for hard links that span repos (and general handling of hard links that cross repo boundaries)
+
+## Open Questions
+
+- During a vnode operation, when a vnode does not exist in the cache (or has been recycled since being added to the cache) should PrjFSKext check `FileFlags_IsInVirtualizationRoot` before it walks up the tree looking for a root?
+- Will PrjFSKext invalidate the cache too frequently if it does so on every directory rename?
+- What is the timeline for allowing breaking changes to existing repos (as we refine the approach that we're taking)?
```

Comment:
I'm unclear how a cache in the kext is a breaking change unless it's persisted somewhere, which I didn't see called out. Can you elaborate?

Reply:
Thinking about it some more, I guess the absence of the file flag will be a breaking change, but initially, I assume we'll still be writing it, it'll just be ignored.

Reply:
Yep it's due to the fact that the cache allows us to change what we're persisting on placeholder files. We're also considering a separate change to store each placeholder file's canonical relative path in its xattr (i.e. the path that we created it with) and avoiding any reliance on resolving paths in the kext.

Breaking changes are a non issue at the moment since we have no active users. We'll definitely need to decide when to set that expectation, but I don't think it needs to be a day 1 requirement necessarily. For reference, we had real users on Windows for at least 6 months before we started guaranteeing that a repo could be used after upgrade. In the beginning, it was more important, and valuable to our users, to be able to settle on the right design without being constrained by prior data shapes. I think we'll settle much more quickly for the Mac client though since the VFSForGit data shapes are already stable, so we just need to let the placeholder file formats stabilize.

Reply:
Yep @sanoursa is correct, and I will remove this question from the document (strictly speaking it's not related to the vnode cache).

The change to remove the flag is not, in and of itself, a breaking change.  However, if we make that change and then discover unforeseen circumstances that require bringing back the flag we'd have a breaking change.

I raised this question in the document because it felt like we're starting down a path that could lead to a breaking change, and I wanted to make sure everyone was on the same page as to how we'd handle that.  I agree with @sanoursa's point that we should make sure we've settled on the right design before locking down the format.